medmodels/data_synthesis/mtgan/modules/preprocessor.py

```diff
@@ -0,0 +1,593 @@
+        for concept_index, concept in enumerate(
+            sorted(medrecord.nodes_in_group(self.concepts_group))
+        ):
+            concepts_list.append(concept)
+            medrecord.node[concept, concept_index_attribute] = concept_index
+
+        return concepts_list
+
+    def preprocess(
+        self, medrecord: MedRecord
+    ) -> Tuple[MedRecord, ConceptsList, PreprocessingAttributes]:
+        """Preprocess the MedRecord.
+
+        - Remove uncommon concepts.
+        - Sample patients.
+        - Find the first admission for each patient.
+        - Remove admissions with less than min_codes_per_window codes.
+        - Find the relative times for each patient per concept.
+
+        Args:
+            medrecord (MedRecord): The MedRecord.
+
+        Returns:
+            Tuple[MedRecord, ConceptsList, PreprocessingAttributes]:
+                The preprocessed MedRecord, the concepts list, and the preprocessing
+                attributes' names in the preprocessed MedRecord.
+
+        Raises:
+            ValueError: If no patients or no concepts are in the MedRecord groups with
+                the given group names before preprocessing. Also another ValueError
+                raised if none are left after preprocessing in either group. If no
+                edges have the given time attribute, also a ValueError is raised.
+
+        Note:
+            It requires the time attribute to be present in the edges of the MedRecord
+                and its values to be datetime objects.
+            This method will remove patient nodes with less than two unique time
+                windows.
+        """
+        if not medrecord.select_edges(edge().has_attribute(self.time_attribute)):
+            raise ValueError("No edges in the MedRecord with that time attribute")
+
+        # TODO: copy of a MedRecord instead of modifying the original one once it is implemented
+        self._remove_uncommon_concepts(
+            medrecord,
+            minimum_number_ocurrences=self.hyperparameters[
+                "minimum_occurrences_concept"
+            ],
+        )
+        concept_index_attribute = self._get_attribute_name(medrecord, "concept_index")
+        first_admission_attribute = self._find_first_admission(medrecord)
+        preprocessing_attributes = self._find_relative_times(
+            medrecord,
+            first_admission_attribute,
+            concept_index_attribute,
+            self.hyperparameters["time_interval_days"],
+            self.hyperparameters["minimum_concepts_per_window"],
+        )
+        # Pruning the concepts after removing some edges and patient nodes
+        self._remove_uncommon_concepts(
+            medrecord,
+            minimum_number_ocurrences=1,
+        )
+        self._sample_patients(
+            medrecord, self.hyperparameters["number_of_sampled_patients"]
+        )
+        concepts_list = self._assign_concept_indices(medrecord, concept_index_attribute)
```

Comment:
Here we just have to be careful to not remove anything again if the concept_index is saved only based on the position in the list.

Reply:
Yes, that is important. Any idea, @JabobKrauskopf ? Or should we live with that?

Reply:
is that a thing that can/should happen? I would have guessed that no concepts are removed once the concept mapping exists.

Reply:
No concepts should be removed, nope.

Reply:
then maybe add a comment that the index of the concept is based on its list index and hope you see it if anything should ever change ^^

Reply:
I realized that having "concept_edge_attribute" erased was a mistake:
- The concept edge attribute holds the NAME of the concept ("Warfarin") in the EDGE connecting that concept to a patient node (easier querying afterwards but probably could be changed when the query engine appears)
- The concept index attribute holds the NUMERICAL INDEX of the concept (0, 1, 2, 3...)  in the concept NODE.

This sounds confusing to me too with the naming and so, so please feel free to suggest better naming.